File: server/services/ExtractionService.ts
import axios from 'axios';
import * as cheerio from 'cheerio';

export class ExtractionService {
  
  static async scrapeWebsite(url: string) {
    try {
      const response = await axios.get(url, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; ModelCardBot/1.0)'
        },
        timeout: 10000
      });

      const $ = cheerio.load(response.data);
      
      // Extract structured data
      const extractedData = {
        title: $('title').text().trim(),
        description: $('meta[name="description"]').attr('content') || '',
        content: this.extractTextContent($),
        metrics: this.extractMetrics($),
        datasetInfo: this.extractDatasetInfo($),
        technicalSpecs: this.extractTechnicalSpecs($),
        sourceUrl: url,
        extractedAt: new Date().toISOString()
      };

      return extractedData;
    } catch (error) {
      console.error('Website scraping failed:', error);
      throw new Error(`Failed to scrape website: ${error.message}`);
    }
  }

  static async extractFromPaper(doi: string) {
    try {
      // Use CrossRef API to get paper metadata
      const crossRefResponse = await axios.get(`https://api.crossref.org/works/${doi}`);
      const paperMetadata = crossRefResponse.data.message;

      // Attempt to get full text from various sources
      const extractedData = {
        title: paperMetadata.title?.[0] || 'Unknown Title',
        authors: paperMetadata.author?.map(a => `${a.given} ${a.family}`) || [],
        journal: paperMetadata['container-title']?.[0] || 'Unknown Journal',
        year: paperMetadata.published?.['date-parts']?.[0]?.[0] || 'Unknown Year',
        doi: doi,
        abstract: paperMetadata.abstract || '',
        metrics: await this.extractPaperMetrics(doi),
        datasetInfo: await this.extractDatasetFromPaper(doi),
        sourceType: 'peer-reviewed-paper',
        extractedAt: new Date().toISOString()
      };

      return extractedData;
    } catch (error) {
      console.error('Paper extraction failed:', error);
      throw new Error(`Failed to extract from paper: ${error.message}`);
    }
  }

  static async extractFromGitHub(repoUrl: string) {
    try {
      const repoPath = repoUrl.replace('https://github.com/', '');
      const apiUrl = `https://api.github.com/repos/${repoPath}`;
      
      // Get repository metadata
      const repoResponse = await axios.get(apiUrl);
      const repoData = repoResponse.data;

      // Get README content
      const readmeResponse = await axios.get(`${apiUrl}/readme`, {
        headers: {
          'Accept': 'application/vnd.github.v3.raw'
        }
      });

      // Extract technical information from README
      const extractedData = {
        name: repoData.name,
        description: repoData.description,
        stars: repoData.stargazers_count,
        language: repoData.language,
        readme: readmeResponse.data,
        metrics: this.extractMetricsFromReadme(readmeResponse.data),
        technicalSpecs: this.extractTechnicalSpecsFromReadme(readmeResponse.data),
        requirements: this.extractRequirements(readmeResponse.data),
        sourceType: 'github-repository',
        sourceUrl: repoUrl,
        extractedAt: new Date().toISOString()
      };

      return extractedData;
    } catch (error) {
      console.error('GitHub extraction failed:', error);
      throw new Error(`Failed to extract from GitHub: ${error.message}`);
    }
  }

  static async extractFromHuggingFace(modelName: string) {
    try {
      const apiUrl = `https://huggingface.co/api/models/${modelName}`;
      
      // Get model metadata
      const modelResponse = await axios.get(apiUrl);
      const modelData = modelResponse.data;

      // Get model card content
      let modelCard = '';
      try {
        const cardResponse = await axios.get(`https://huggingface.co/${modelName}/raw/main/README.md`);
        modelCard = cardResponse.data;
      } catch (e) {
        console.warn('Could not fetch model card README');
      }

      const extractedData = {
        name: modelData.modelId || modelName,
        downloads: modelData.downloads || 0,
        likes: modelData.likes || 0,
        tags: modelData.tags || [],
        pipeline: modelData.pipeline_tag || '',
        modelCard: modelCard,
        metrics: this.extractMetricsFromModelCard(modelCard),
        datasets: this.extractDatasetsFromModelCard(modelCard),
        sourceType: 'huggingface-model',
        sourceUrl: `https://huggingface.co/${modelName}`,
        extractedAt: new Date().toISOString()
      };

      return extractedData;
    } catch (error) {
      console.error('HuggingFace extraction failed:', error);
      throw new Error(`Failed to extract from HuggingFace: ${error.message}`);
    }
  }

  // Helper methods for parsing content
  private static extractTextContent($: cheerio.CheerioAPI): string {
    // Remove script and style elements
    $('script, style').remove();
    
    // Extract main content areas
    const contentSelectors = [
      'main', 'article', '.content', '#content', 
      '.documentation', '.model-info', '.performance'
    ];
    
    let content = '';
    for (const selector of contentSelectors) {
      const element = $(selector);
      if (element.length > 0) {
        content += element.text().trim() + ' ';
      }
    }
    
    return content || $('body').text().trim();
  }

  private static extractMetrics($: cheerio.CheerioAPI): any {
    const metrics: any = {};
    
    // Look for common metric patterns
    const text = $.text().toLowerCase();
    
    // Extract accuracy
    const accuracyMatch = text.match(/accuracy[:\s]*([0-9.]+)%?/i);
    if (accuracyMatch) {
      metrics.accuracy = parseFloat(accuracyMatch[1]) / (accuracyMatch[1].includes('.') && parseFloat(accuracyMatch[1]) <= 1 ? 1 : 100);
    }
    
    // Extract sensitivity/recall
    const sensitivityMatch = text.match(/(?:sensitivity|recall)[:\s]*([0-9.]+)%?/i);
    if (sensitivityMatch) {
      metrics.sensitivity = parseFloat(sensitivityMatch[1]) / (sensitivityMatch[1].includes('.') && parseFloat(sensitivityMatch[1]) <= 1 ? 1 : 100);
    }
    
    // Extract specificity
    const specificityMatch = text.match(/specificity[:\s]*([0-9.]+)%?/i);
    if (specificityMatch) {
      metrics.specificity = parseFloat(specificityMatch[1]) / (specificityMatch[1].includes('.') && parseFloat(specificityMatch[1]) <= 1 ? 1 : 100);
    }
    
    // Extract AUC
    const aucMatch = text.match(/auc[:\s]*([0-9.]+)/i);
    if (aucMatch) {
      metrics.auc = parseFloat(aucMatch[1]);
    }

    return Object.keys(metrics).length > 0 ? metrics : null;
  }

  private static extractDatasetInfo($: cheerio.CheerioAPI): any {
    const text = $.text().toLowerCase();
    const datasetInfo: any = {};
    
    // Extract dataset size
    const sizeMatch = text.match(/(?:dataset|training)[^0-9]*([0-9,]+)[^0-9]*(?:images|samples|patients)/i);
    if (sizeMatch) {
      datasetInfo.size = sizeMatch[1].replace(',', '');
    }
    
    // Extract dataset names
    const commonDatasets = ['isic', 'dermnet', 'ham10000', 'asan', 'fitzpatrick'];
    for (const dataset of commonDatasets) {
      if (text.includes(dataset)) {
        datasetInfo.name = dataset.toUpperCase();
        break;
      }
    }
    
    return Object.keys(datasetInfo).length > 0 ? datasetInfo : null;
  }

  private static extractTechnicalSpecs($: cheerio.CheerioAPI): any {
    const text = $.text().toLowerCase();
    const specs: any = {};
    
    // Extract model architecture
    const architectures = ['resnet', 'vgg', 'densenet', 'efficientnet', 'mobilenet', 'inception'];
    for (const arch of architectures) {
      if (text.includes(arch)) {
        specs.architecture = arch;
        break;
      }
    }
    
    // Extract framework
    const frameworks = ['pytorch', 'tensorflow', 'keras', 'scikit-learn'];
    for (const framework of frameworks) {
      if (text.includes(framework)) {
        specs.framework = framework;
        break;
      }
    }
    
    return Object.keys(specs).length > 0 ? specs : null;
  }

  private static async extractPaperMetrics(doi: string): Promise<any> {
    // In production, this would parse PDFs or use academic APIs
    // For now, return placeholder that indicates real extraction attempt
    return {
      extractionNote: `Attempted real extraction from DOI: ${doi}`,
      extractionStatus: 'requires-pdf-access'
    };
  }

  private static async extractDatasetFromPaper(doi: string): Promise<any> {
    // Would extract dataset information from paper content
    return {
      extractionNote: `Dataset info would be extracted from paper: ${doi}`,
      extractionStatus: 'requires-full-text-access'
    };
  }

  private static extractMetricsFromReadme(readme: string): any {
    const metrics: any = {};
    const text = readme.toLowerCase();
    
    // Extract metrics from README
    const accuracyMatch = text.match(/accuracy[:\s]*([0-9.]+)%?/i);
    if (accuracyMatch) {
      metrics.accuracy = parseFloat(accuracyMatch[1]);
      metrics.source = 'README.md';
    }
    
    return Object.keys(metrics).length > 0 ? metrics : null;
  }

  private static extractTechnicalSpecsFromReadme(readme: string): any {
    const specs: any = {};
    const text = readme.toLowerCase();
    
    // Extract technical specifications
    if (text.includes('pytorch')) specs.framework = 'PyTorch';
    if (text.includes('tensorflow')) specs.framework = 'TensorFlow';
    
    return Object.keys(specs).length > 0 ? specs : null;
  }

  private static extractRequirements(readme: string): string[] {
    const requirements = [];
    const lines = readme.split('\n');
    
    for (const line of lines) {
      if (line.toLowerCase().includes('requirement') || line.toLowerCase().includes('install')) {
        const pythonPackages = line.match(/pip install ([^\n]+)/);
        if (pythonPackages) {
          requirements.push(...pythonPackages[1].split(' '));
        }
      }
    }
    
    return requirements;
  }

  private static extractMetricsFromModelCard(modelCard: string): any {
    const metrics: any = {};
    const text = modelCard.toLowerCase();
    
    // Extract metrics from model card
    const accuracyMatch = text.match(/accuracy[:\s]*([0-9.]+)%?/);
    if (accuracyMatch) {
      metrics.accuracy = parseFloat(accuracyMatch[1]);
      metrics.source = 'HuggingFace Model Card';
    }
    
    return Object.keys(metrics).length > 0 ? metrics : null;
  }

  private static extractDatasetsFromModelCard(modelCard: string): string[] {
    const datasets = [];
    const text = modelCard.toLowerCase();
    
    const commonDatasets = ['imagenet', 'coco', 'isic', 'dermnet', 'ham10000'];
    for (const dataset of commonDatasets) {
      if (text.includes(dataset)) {
        datasets.push(dataset);
      }
    }
    
    return datasets;
  }
}

File: python_app/services/extraction_service.py
import httpx
from bs4 import BeautifulSoup
import re
from typing import Dict, Any, List, Optional
from datetime import datetime

from models import ExtractedData, PaperData, GitHubData


class ExtractionService:
    """Service for extracting data from various sources"""
    
    @staticmethod
    async def scrape_website(url: str) -> ExtractedData:
        """Scrape and extract data from a website"""
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                headers = {
                    'User-Agent': 'Mozilla/5.0 (compatible; ModelCardBot/1.0)'
                }
                response = await client.get(url, headers=headers)
                response.raise_for_status()
            
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # Extract structured data
            title = soup.find('title')
            title_text = title.text.strip() if title else ''
            
            description_meta = soup.find('meta', attrs={'name': 'description'})
            description = description_meta.get('content', '') if description_meta else ''
            
            extracted_data = ExtractedData(
                title=title_text,
                description=description,
                content=ExtractionService._extract_text_content(soup),
                metrics=ExtractionService._extract_metrics(soup),
                datasetInfo=ExtractionService._extract_dataset_info(soup),
                technicalSpecs=ExtractionService._extract_technical_specs(soup),
                sourceUrl=url,
                sourceType="website"
            )
            
            return extracted_data
            
        except Exception as e:
            raise Exception(f"Failed to scrape website: {str(e)}")
    
    @staticmethod
    async def extract_from_paper(doi: str) -> PaperData:
        """Extract data from academic paper using DOI"""
        try:
            # Use CrossRef API to get paper metadata
            async with httpx.AsyncClient() as client:
                crossref_url = f"https://api.crossref.org/works/{doi}"
                response = await client.get(crossref_url)
                response.raise_for_status()
                
                data = response.json()
                paper_metadata = data['message']
            
            # Extract authors
            authors = []
            if 'author' in paper_metadata:
                for author in paper_metadata['author']:
                    given = author.get('given', '')
                    family = author.get('family', '')
                    if given and family:
                        authors.append(f"{given} {family}")
            
            # Extract publication year
            year = 'Unknown Year'
            if 'published' in paper_metadata and 'date-parts' in paper_metadata['published']:
                date_parts = paper_metadata['published']['date-parts']
                if date_parts and len(date_parts[0]) > 0:
                    year = str(date_parts[0][0])
            
            paper_data = PaperData(
                title=paper_metadata.get('title', ['Unknown Title'])[0],
                authors=authors,
                journal=paper_metadata.get('container-title', ['Unknown Journal'])[0],
                year=year,
                doi=doi,
                abstract=paper_metadata.get('abstract', ''),
                metrics=await ExtractionService._extract_paper_metrics(doi),
                datasetInfo=await ExtractionService._extract_dataset_from_paper(doi)
            )
            
            return paper_data
            
        except Exception as e:
            raise Exception(f"Failed to extract from paper: {str(e)}")
    
    @staticmethod
    async def extract_from_github(repo_url: str) -> GitHubData:
        """Extract data from GitHub repository"""
        try:
            # Convert GitHub URL to API format
            repo_path = repo_url.replace('https://github.com/', '')
            api_url = f"https://api.github.com/repos/{repo_path}"
            
            async with httpx.AsyncClient() as client:
                # Get repository metadata
                repo_response = await client.get(api_url)
                repo_response.raise_for_status()
                repo_data = repo_response.json()
                
                # Get README content
                readme_url = f"{api_url}/readme"
                readme_headers = {'Accept': 'application/vnd.github.v3.raw'}
                readme_response = await client.get(readme_url, headers=readme_headers)
                readme_content = readme_response.text if readme_response.status_code == 200 else ''
            
            github_data = GitHubData(
                name=repo_data['name'],
                description=repo_data.get('description', ''),
                stars=repo_data['stargazers_count'],
                language=repo_data.get('language'),
                readme=readme_content,
                metrics=ExtractionService._extract_metrics_from_readme(readme_content),
                technicalSpecs=ExtractionService._extract_technical_specs_from_readme(readme_content),
                requirements=ExtractionService._extract_requirements(readme_content),
                sourceUrl=repo_url
            )
            
            return github_data
            
        except Exception as e:
            raise Exception(f"Failed to extract from GitHub: {str(e)}")
    
    @staticmethod
    def _extract_text_content(soup: BeautifulSoup) -> str:
        """Extract main text content from HTML"""
        # Remove script and style elements
        for script in soup(["script", "style"]):
            script.decompose()
        
        # Get text content
        text = soup.get_text()
        lines = (line.strip() for line in text.splitlines())
        chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
        text = ' '.join(chunk for chunk in chunks if chunk)
        
        return text[:5000]  # Limit to first 5000 characters
    
    @staticmethod
    def _extract_metrics(soup: BeautifulSoup) -> Dict[str, Any]:
        """Extract performance metrics from HTML"""
        metrics = {}
        
        # Look for common metric patterns
        text = soup.get_text().lower()
        
        # Accuracy patterns
        accuracy_match = re.search(r'accuracy[:\s]*(\d+(?:\.\d+)?%?)', text)
        if accuracy_match:
            metrics['accuracy'] = accuracy_match.group(1)
        
        # AUC patterns
        auc_match = re.search(r'auc[:\s]*(\d+(?:\.\d+)?)', text)
        if auc_match:
            metrics['auc'] = float(auc_match.group(1))
        
        # Sensitivity/Specificity patterns
        sensitivity_match = re.search(r'sensitivity[:\s]*(\d+(?:\.\d+)?%?)', text)
        if sensitivity_match:
            metrics['sensitivity'] = sensitivity_match.group(1)
        
        specificity_match = re.search(r'specificity[:\s]*(\d+(?:\.\d+)?%?)', text)
        if specificity_match:
            metrics['specificity'] = specificity_match.group(1)
        
        return metrics
    
    @staticmethod
    def _extract_dataset_info(soup: BeautifulSoup) -> Dict[str, Any]:
        """Extract dataset information from HTML"""
        dataset_info = {}
        text = soup.get_text().lower()
        
        # Look for dataset size
        size_match = re.search(r'(\d+(?:,\d+)*)\s*(?:images|samples|cases)', text)
        if size_match:
            dataset_info['size'] = size_match.group(1)
        
        # Look for dataset name
        dataset_patterns = [
            r'dataset[:\s]*([a-zA-Z0-9\-_\s]+)',
            r'trained on[:\s]*([a-zA-Z0-9\-_\s]+)',
            r'using[:\s]*([a-zA-Z0-9\-_\s]+)\s*dataset'
        ]
        
        for pattern in dataset_patterns:
            match = re.search(pattern, text)
            if match:
                dataset_info['name'] = match.group(1).strip()
                break
        
        return dataset_info
    
    @staticmethod
    def _extract_technical_specs(soup: BeautifulSoup) -> Dict[str, Any]:
        """Extract technical specifications from HTML"""
        specs = {}
        text = soup.get_text().lower()
        
        # Look for model architecture
        arch_patterns = ['resnet', 'densenet', 'efficientnet', 'inception', 'vgg', 'mobilenet']
        for arch in arch_patterns:
            if arch in text:
                specs['architecture'] = arch
                break
        
        # Look for framework
        frameworks = ['tensorflow', 'pytorch', 'keras', 'scikit-learn']
        for framework in frameworks:
            if framework in text:
                specs['framework'] = framework
                break
        
        return specs
    
    @staticmethod
    async def _extract_paper_metrics(doi: str) -> Dict[str, Any]:
        """Extract metrics from research paper"""
        # This would typically involve more sophisticated NLP
        # For now, return empty dict
        return {}
    
    @staticmethod
    async def _extract_dataset_from_paper(doi: str) -> Dict[str, Any]:
        """Extract dataset information from research paper"""
        # This would typically involve more sophisticated NLP
        # For now, return empty dict
        return {}
    
    @staticmethod
    def _extract_metrics_from_readme(readme_content: str) -> Dict[str, Any]:
        """Extract metrics from README content"""
        metrics = {}
        text = readme_content.lower()
        
        # Look for common metric patterns in README
        accuracy_match = re.search(r'accuracy[:\s]*(\d+(?:\.\d+)?%?)', text)
        if accuracy_match:
            metrics['accuracy'] = accuracy_match.group(1)
        
        return metrics
    
    @staticmethod
    def _extract_technical_specs_from_readme(readme_content: str) -> Dict[str, Any]:
        """Extract technical specs from README content"""
        specs = {}
        text = readme_content.lower()
        
        # Look for Python version
        python_match = re.search(r'python\s*(\d+\.\d+)', text)
        if python_match:
            specs['python_version'] = python_match.group(1)
        
        return specs
    
    @staticmethod
    def _extract_requirements(readme_content: str) -> List[str]:
        """Extract requirements from README content"""
        requirements = []
        
        # Look for requirements section
        lines = readme_content.split('\n')
        in_requirements = False
        
        for line in lines:
            if 'requirements' in line.lower() or 'dependencies' in line.lower():
                in_requirements = True
                continue
            
            if in_requirements:
                if line.strip().startswith('- ') or line.strip().startswith('* '):
                    req = line.strip()[2:].strip()
                    if req:
                        requirements.append(req)
                elif line.strip() == '':
                    continue
                else:
                    break
        
        return requirements